### PR TITLE
dvc.api for dataset

### LINKED ISSUE #10313: `dvc.api.dataset`
API for new `datasets` feature

- [ ] DVCX (highest priority) - return name, version
- [ ] DVC (next priority) - return URL, path, sha
- [ ] Cloud-versioned (nice to have) - return URL, version ID for each file

Needs spec

### PR DESCRIPTION
**Pull request by skshetry**
On top of #10287. Closes #10313.

### Usage
```python
from dvc.api import get_dataset

get_dataset("<dataset_name>")
```
The `get_dataset()` API returns a dict with values depending on the type of dataset.

### dvcx
```python
from dvc.api import get_dataset
from dvcx.query import DatasetQuery

d = get_dataset("dogs")
print(d)

assert d["type"] == "dvcx"
query = DatasetQuery(name=d["name"], version=d["version"])
```

```python
{
    'type': 'dvcx',
    'name': 'dogs',
    'version': 3,
}
```

### dvc
```python
from dvc.api import DVCFileSystem, get_dataset

d = get_dataset("example-get-started")
print(d)

assert d["type"] == "dvc"
fs = DVCFileSystem(d["url"], rev=d["sha"])
fs.read_text(d["path"])
```

```python
{
    'type': 'dvc',
    'url': 'git@github.com:iterative/example-get-started.git',
    'path': 'path',
    'sha': 'df75c16ef61f0772d6e4bb27ba4617b06b4b5398'
}
```

### cloud-versioned and non-cloud-versioned remotes
```python
from dvc.api import get_dataset
from fsspec import filesystem

d = get_dataset("cloud-demo")
fs = filesystem("s3")

assert d["type"] == "url"
fs.cat_file(d["path"]) # for a single file
fs.cat_file(d["files"][0]) # for a _directory_
```

```python
{
    'type': 'url',
    'files': [
        's3://cloud-versioning-demo/myproject/model.pt?versionId=5qrtnhnQ4fBzV73kqqK6pMGhTOzd_IPr',
        's3://cloud-versioning-demo/remote/predictions/1.txt?versionId=M2CEHoqXy3JZq07EmiLRMQCuj0wU8mbp'
    ],
    'path': 's3://cloud-versioning-demo'
}
```

> [!NOTE]
> This will also expand `remote://` URLs and return a canonicalized path.

---

The type is a union of the following typeddict:
```python
class DVCXDataset(TypedDict):
    type: Literal["dvcx"]
    name: str
    version: int


class DVCDataset(TypedDict):
    type: Literal["dvc"]
    url: str
    path: str
    sha: str


class URLDataset(TypedDict):
    type: Literal["url"]
    files: list[str]
    path: str
```

Users can narrow type with `assert d["type"] == "dvcx"` (the sole reason why `type` key exists).

### CODE REVIEW

File: dvc/api/dataset.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    type: Literal["dvcx"]
+    name: str
+    version: int
+
+
+class DVCDataset(TypedDict):
+    type: Literal["dvc"]
+    url: str
+    path: str
+    sha: str
```
Comment (dberenbaum) [inline dvc/api/dataset.py:14]: Minor, but can we name it `rev` to match the rest of the API?

Comment (skshetry) [reply dvc/api/dataset.py:14]: This is not a rev, but actually a `sha`/`commit-hash`.

**Comment by dberenbaum** [comment]
> from dvc.api.dataset import get

Do you anticipate additional methods? Even if you want a separate module, should we just expose it as a top-level API method like `get_ds()`?

**Comment by skshetry** [comment]
> > from dvc.api.dataset import get
> 
> Do you anticipate additional methods? Even if you want a separate module, should we just expose it as a top-level API method like `get_ds()`?

I can rename it to `get_dataset` under top-level.

**Review by dberenbaum — APPROVED** [review]

**Pull request merged**